cpp/include/cudf/detail/iterator.cuh

```diff
@@ -153,6 +183,54 @@ auto inline make_validity_iterator(column_device_view const& column)
       validity_accessor{column});
 }
 
+/**
+ * @brief Constructs a constant iterator over a scalar's value.
+ *
+ * Dereferencing the returned iterator returns a `Element`.
+ * 
+ * For `p = *(iter + i)`, `p` is the value stored in the scalar.
+ *
+ * @throws `cudf::logic_error` if scalar datatype and Element type mismatch.
+ * @throws `cudf::logic_error` if scalar is null.
+ *
+ * @tparam Element The type of elements in the scalar
+ * @param scalar_value The scalar to iterate
+ * @return auto Iterator that returns scalar value
+ */
+template <typename Element>
+auto inline make_scalar_iterator(scalar const& scalar_value)
+{
+  CUDF_EXPECTS(data_type(experimental::type_to_id<Element>()) == scalar_value.type(), "the data type mismatch");
+  CUDF_EXPECTS(scalar_value.is_valid(), "the scalar value must be valid");
+  using ScalarType = experimental::scalar_type_t<Element>;
+  return thrust::make_constant_iterator(static_cast<ScalarType const*>(&scalar_value)->value());
+}
+
+/**
+ * @brief Constructs a constant pair iterator over a scalar's value and its validity.
+ *
+ * Dereferencing the returned iterator returns a `thrust::pair<Element, bool>`.
+ *  
+ * If scalar is valid, then for `p = *(iter + i)`, `p.first` contains 
+ * the value of the scalar and `p.second == true`.
+ *
+ * Else, if the scalar is null, then the value of `p.first` is undefined and `p.second == false`. 
+ * 
+ * @throws `cudf::logic_error` if scalar datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the scalar
+ * @tparam is_null boolean unused
+ * @param scalar_value The scalar to iterate
+ * @return auto Iterator that returns scalar, and validity of the scalar in a pair
+ */
+template <typename Element, bool is_null=false>
```

Comment:
The `is_null` value should be checked against `scalar_value.is_valid()`.

Reply:
`is_null` tparam is added just to keep similar interface as `make_pair_iterator(column_device_view)` as a requirement https://github.com/rapidsai/cudf/pull/3490#discussion_r362013093. It is not supposed to be used.

Reply:
I removed the name `is_null` to avoid confusion.

Reply:
Couple of comments:

- I like the idea of having the parameter actually being named.  It actually serves a real purpose, so the user should know it's there.

- Related, I think it probably shouldn't have a default value, similar to the column_device_view.  The danger there is the user actually has a scalar that's null, but is quietly wrapping as if it weren't.

- I agree with Jake on verifying the template parameter against the actual is_null() of the scalar at construction time. 

- It seems like a good idea to use the template parameter to stuff the validity value in the returned pairs instead of always calling is_null().

Reply:
@nvdbaranec 
`has_nulls` tparam for `make_pair_iterator(column_device_view)` is present for 2 purposes - 
1. prevent accidentally access invalid null_mask ptr. 
2. run time speed up we could get for iterators by doing compile time optimization for non-nullable columns. 

In `make_pair_iterator(scalar)`, both purposes are not required. `is_valid` always is stored in `scalar` and `is_valid()` value is always returned for pair iterator. The user is not quietly wrapping it because we return a pair(value, validity).

In fact, adding `is_null==is_valid()` check does additional data transfer to host. (`is_null` is different from `has_nulls`).

Reply:
Ah, you're right.  This would only make sense if the scalar kept a cpu-side cached version of _is_valid around.  Then the user could cheaply do an if() to create a <true> or <false> pair iterator.  This is kind of like the cached null count in a column_view and calling has_nulls().